designs/generics-design.md

```diff
@@ -0,0 +1,403 @@
+abstract over the specific types of data that they work with. This is done by
+parameterizing classes and functions using type variables and type parameters.
+
+1. **Basic Type Variable Declaration**
+
+```python
+T = TypeVar('T')
+```
+
+A type variable can be introduced using a TypeVar definition. New type variables
+can be introduced only in global scope and can be used across any number of generic
+class/function definitions.
+
+The first parameter to TypeVar can be used to provide a more descriptive name
+for the type variable and can potentially be used for better error
+messages from the compiler for example.
+
+2. **Class Definition with a Type Variable**
+
+```python
+T = TypeVar('T')
+
+class Box(Generic[T]):
+    pass
+```
+
+A generic class can be defined by inheriting from the special base class Generic
+along with a defined type variable. Any other class that this class
+inherits from in the inheritance hierarchy can be specified after this.
+
+A generic class is type-checked for all possible allowed instantiations of its
+type variables independent of the instantiations actually made in the program.
+
+3. **Object creation using Type Variable instantiated with a Primitive Type**
+```python
+T = TypeVar('T')
+
+class Box(Generic[T]):
+    pass
+
+b: Box[int] = None
+b = Box()
+```
+
+An object of a generic class can be created by instantiating the type parameter
+with a primitive type in the type annotation for the object's variable declaration.
+
+After type-checking and before lowering a copy of each generic class is made for
+each type that its type variables are instantiated with over the course of the program.
+
+
+4. **Object creation using Type Variable instantiated with a User-Defined Type**
+
+```python
+T = TypeVar('T')
+
+class Box(Generic[T]):
+    pass
+
+class Rat():
+    pass
+
+b: Box[Rat] = None
+b = Box()
+```
+
+A type parameter of a generic class can also be instantiated with a user-defined
+class.
+
+5. **Type Variables and Generic Classes with Constraints**
+
+```python
+T = TypeVar('T')
+R = TypeVar('R', int, bool)
+
+class Box(Generic[T]):
+    pass
+
+class ConstrainedBox(Generic[R]):
+    pass
+
+b1: Box[int] = Box()
+cb1: ConstrainedBox[bool] = ConstrainedBox()
+```
+
+A type variable definition may include a list of types that constrains the types
+that the variable can be instantiated with. The list of types are interpreted as
+"OR" constraints on the type variable.
+
+6. **Generic Class with Fields of a Generic Type**
+```python
+T = TypeVar('T')
+
+class Box(Generic[T]):
+    f : T = __ZERO__
+
+b : Box[int] = None
+b = Box()
+print(b.f) #prints 0 (zero value of int)
+b.f = 10
+print(b.f) #prints 10
+```
+
+Generic classes can have fields with the type parameter as a type annotation. Since
+fields need to always be initialized but we don't know the actual type of the field, generic
+fields need to be initialized with the special __ZERO__ literal value that automatically uses
```

Comment:
Ah fascinating! Nice to leave this until later. Special symbols is an interesting idea. I wonder if we need a better convention for them, but this is fine for a start.